L1Trigger/L1TTrackMatch/src/L1TkMuCorrDynamicWindows.cc

```diff
@@ -0,0 +1,401 @@
+    float trk_theta = to_mpio2_pio2(eta_to_theta(l1trkit->momentum().eta()));
+    float trk_phi = l1trkit->momentum().phi();
+    int trk_charge = (l1trkit->rInv() > 0 ? 1 : -1);
+
+    // porting some selections from the MuonTrackCorr finder
+    // https://github.com/cms-l1t-offline/cmssw/blob/l1t-phase2-932-v1.6/L1Trigger/L1TTrackMatch/plugins/L1TkMuonProducer.cc#L264
+    // in future, make preselections confiuguable
+    bool reject_trk = false;
+    if (trk_p < min_trk_p_)
+      reject_trk = true;
+    if (trk_aeta > max_trk_aeta_)
+      reject_trk = true;
+    if (track_qual_presel_) {
+      float l1tk_chi2 = l1trkit->chi2();
+      int l1tk_nstubs = l1trkit->getStubRefs().size();
+      if (l1tk_chi2 >= max_trk_chi2_)
+        reject_trk = true;
+      if (l1tk_nstubs < min_trk_nstubs_)
+        reject_trk = true;
+    }
+
+    int ibin = findBin(trk_aeta);
+
+    std::vector<std::tuple<float, float, int>> matched;  // dtheta, dphi, idx
+    // loop on muons to see which match
+    for (auto l1muit = l1mus.begin(); l1muit != l1mus.end(); ++l1muit) {
+      // match only muons in the central bx - as the track collection refers anyway to bx 0 only
+      if (l1muit->BX() != 0)
+        continue;
+
+      // putting everything in rad
+      float emtf_theta = to_mpio2_pio2(eta_to_theta(l1muit->Eta()));
+      float emtf_phi = deg_to_rad(l1muit->Phi_glob());
+
+      float dtheta = std::abs(emtf_theta - trk_theta);
+      float dphi = to_mpi_pi(emtf_phi - trk_phi);
+      float adphi = std::abs(dphi);
+
+      double sf_l;
+      double sf_h;
+      if (do_relax_factor_) {
+        sf_l = sf_progressive(trk_pt, pt_start_, pt_end_, initial_sf_l_, safety_factor_l_);
+        sf_h = sf_progressive(trk_pt, pt_start_, pt_end_, initial_sf_h_, safety_factor_h_);
+      } else {
+        sf_l = safety_factor_l_;
+        sf_h = safety_factor_h_;
+      }
+
+      if (
+          // emtf_theta * trk_theta > 0 &&
+          dtheta > (1 - sf_l) * wdws_theta_.at(ibin).bound_low(trk_pt) &&
+          dtheta <= (1 + sf_h) * wdws_theta_.at(ibin).bound_high(trk_pt) &&
+          adphi > (1 - sf_l) * wdws_phi_.at(ibin).bound_low(trk_pt) &&
+          adphi <= (1 + sf_h) * wdws_phi_.at(ibin).bound_high(trk_pt) && dphi * trk_charge < 0 &&  // sign requirement
+          // rndm > 0.5
+          true)
+        matched.push_back(std::make_tuple(dtheta, adphi, std::distance(l1mus.begin(), l1muit)));
+    }
+
+    if (reject_trk)
+      matched.clear();  // quick fix - to be optimised to avoid the operations above
+
+    if (matched.empty())
+      out.at(std::distance(l1trks.begin(), l1trkit)) = -1;
+    else {
+      std::sort(matched.begin(), matched.end());  // closest in theta, then in phi
+      out.at(std::distance(l1trks.begin(), l1trkit)) = std::get<2>(matched.at(0));
+    }
+  }
+
+  // now convert out to a unique set
+  // auto unique_out = make_unique_coll(mtkt, out, narbitrated);
+  auto unique_out = make_unique_coll(l1mus.size(), l1trks, out);
```

Comment:
just return the function output directly to avoid temporary

Reply:
done